lib/tchannel-alpha4.js

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var recvPingReq = require('./swim/ping-req-recvr.js');
+var safeParse = require('./util').safeParse;
+
+var commands = {
+    '/health': 'health',
+
+    '/admin/stats': 'adminStats',
+    '/admin/debugSet': 'adminDebugSet',
+    '/admin/debugClear': 'adminDebugClear',
+    '/admin/gossip': 'adminGossip',
+    '/admin/leave': 'adminLeave',
+    '/admin/join': 'adminJoin',
+    '/admin/reload': 'adminReload',
+    '/admin/tick': 'adminTick',
+
+    '/protocol/leave': 'protocolLeave',
+    '/protocol/join': 'protocolJoin',
+    '/protocol/ping': 'protocolPing',
+    '/protocol/ping-req': 'protocolPingReq',
+
+    '/proxy/req': 'proxyReq'
+};
+
+function RingPopTChannel(ringpop) {
+    this.ringpop = ringpop;
+
+    var self = this;
+
+    return function handleRequest(req, res) {
+        var endpoint = String(req.arg1);
+
+        var fn = self[endpoint];
+        if (!fn) {
+            return false;
+        }
+
+        fn.call(self, req.arg2, req.arg3, req.remoteAddr, onResponse);
+        return true;
+
+        function onResponse(err, res1, res2) {
+            if (err) {
+                res.sendNotOk(null, err.message);
```

Comment:
sendNotOk() will do a call response where `resp.ok === false`. We put the message as a string in arg3